# Lab 2

In [1]:
#pip install IP2Location

In [2]:
import os
import sys
spark_path = os.environ['SPARK_HOME']
sys.path.append(spark_path + "/bin")
sys.path.append(spark_path + "/python")
sys.path.append(spark_path + "/python/pyspark/")
sys.path.append(spark_path + "/python/lib")
sys.path.append(spark_path + "/python/lib/pyspark.zip")
sys.path.append(spark_path + "/python/lib/py4j-0.10.9-src.zip")

import findspark
findspark.init()
import pyspark

In [3]:
number_cores = 6
memory_gb = 16
conf = (pyspark.SparkConf().setMaster('local[{}]'.format(number_cores)).set('spark.driver.memory', '{}g'.format(memory_gb)))
sc = pyspark.SparkContext(conf=conf)

In [4]:
!ls /users/trush/CSC496/PageRank/Lab2

data  IP-COUNTRY.BIN  Lab2.ipynb


In [5]:
auth_log = sc.textFile("/users/trush/CSC496/PageRank/Lab2/data")
auth_log.take(5)

['Nov 15 00:07:01 submitty CRON[31671]: pam_unix(cron:session): session opened for user submitty_daemon by (uid=0)',
 'Nov 15 00:07:01 submitty CRON[31671]: pam_unix(cron:session): session closed for user submitty_daemon',
 'Nov 15 00:07:21 submitty sshd[31682]: pam_unix(sshd:auth): authentication failure; logname= uid=0 euid=0 tty=ssh ruser= rhost=218.92.0.248  user=root',
 'Nov 15 00:07:23 submitty sshd[31682]: Failed password for root from 218.92.0.248 port 11673 ssh2',
 'Nov 15 00:07:27 submitty sshd[31682]: Failed password for root from 218.92.0.248 port 11673 ssh2']

## 1. How many failed attempts to access the server as root are there? List all countries from which these attempts where carried out.

In [17]:
'''
# Total failed attempts to login
total_attempts = auth_log.count()
print(total_attempts)
total_successful_attempts = auth_log.filter(lambda x: "CRON" in x)
print(total_successful_attempts.count())
total_failed_attempts = auth_log.filter(lambda x: "sshd" in x)
print(total_failed_attempts.count())

other_attempts = auth_log.filter(lambda x: "sshd" not in x and "CRON" not in x)
other_attempts.take(50)
'''
# Number of failed attempts to login as root
failed_attempts = auth_log.filter(lambda x: "root" in x and "sshd" in x)
#failed_attempts.take(5)
print(failed_attempts.count())

75630


In [7]:
import IP2Location
import re

database = IP2Location.IP2Location(os.path.join("data", "/users/trush/CSC496/PageRank/Lab2/IP-COUNTRY.BIN"))

def get_country_name(x):
    ip = re.split(pattern=r'(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})', string=x)[1]
    rec = database.get_all(x).country_long
    #print(type(rec))
    return rec

#get_country_name('218.92.0.248')

In [8]:
failed_root_countries = failed_attempts.map(lambda line: get_country_name(line))
#failed_root_countries.take(5)
failed_root_countries.distinct().collect()

Traceback (most recent call last):
  File "/opt/spark-3.0.1-bin-hadoop3.2/python/pyspark/serializers.py", line 468, in dumps
    return cloudpickle.dumps(obj, pickle_protocol)
  File "/opt/spark-3.0.1-bin-hadoop3.2/python/pyspark/cloudpickle.py", line 1097, in dumps
    cp.dump(obj)
  File "/opt/spark-3.0.1-bin-hadoop3.2/python/pyspark/cloudpickle.py", line 357, in dump
    return Pickler.dump(self, obj)
  File "/usr/lib/python3.6/pickle.py", line 409, in dump
    self.save(obj)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 751, in save_tuple
    save(element)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/opt/spark-3.0.1-bin-hadoop3.2/python/pyspark/cloudpickle.py", line 501, in save_function
    self.save_function_tuple(obj)
  File "/opt/spark-3.0.1-bin-hadoop3.2/python/pyspark/cloudpickle.py", li

PicklingError: Could not serialize object: TypeError: cannot serialize '_io.BufferedReader' object

There are a total of 75,630 failed attempts to access the server as root. A filter on the log files returns all entries with attempts to login as root (explain sshd). The function get_country_name splits each line in failed attempts to isolate the IP address and uses the IP2Location database to identify the country location. Then, failed_root_countries is assigned to a map of each country relative to each failed attempt's IP address. Finally, the distinct country names are outputted to identify all countries from which failed attempts were carried out.

## 2. How many failed attempts to access the server as non-root users are there? What are the attempted usernames? List all countries from which these attempts where carried out. 

In [19]:
# Number of failed attempts to login as non-root
failed_nonroot = auth_log.filter(lambda x: "root" not in x and "sshd" in x and "user unknown" not in x)
#failed_nonroot.take(5)
print(failed_nonroot.count())

76541


In [36]:
def get_username(x):
    line = re.split(pattern=r'user ', string=x)[1]
    line = re.split(pattern=r' from', string=line)[0]
    line = re.split(pattern=r'(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})', string=line)[0]
    return line

In [37]:
failed_usernames = failed_nonroot.filter(lambda x: 'user ' in x)
#failed_usernames.take(5)
failed_nonroot_usernames = failed_usernames.map(lambda line: get_username(line))
#failed_nonroot_usernames.take(50)
failed_nonroot_usernames.distinct().collect()

['syslog ',
 'terminal ',
 'boris ',
 'odoo',
 'test',
 'wt',
 'damian ',
 'storm',
 'storm ',
 'jessica',
 'ruben ',
 '',
 'ronald',
 'minecraft',
 'produccion ',
 'test ',
 'pi',
 'test3',
 'git ',
 'ubuntu ',
 'personal',
 'william',
 'designer ',
 'frappe ',
 'user6 ',
 'ppldtepe ',
 'aaron',
 'share',
 'demo ',
 'tomcat ',
 'jimmy',
 'jimmy ',
 'mexal',
 'ark ',
 'tibero6',
 'wang',
 'joshua ',
 'Administrator',
 'support',
 'support ',
 'webapp',
 'tuxedo',
 'backuppc',
 'marek ',
 'king',
 'chrome',
 'zabbix',
 'pydio ',
 'testing',
 'python',
 'fernando',
 'test001',
 'test001 ',
 'ab ',
 'administrador',
 'ts3 ',
 'teamspeak ',
 'server&2008 ',
 '!Z1qaz1qaz',
 '1q@@w3e$$r',
 'power2008',
 '@dMIN123',
 'qwerty!.',
 '121233 ',
 'Pa55w0rD! ',
 'Paradox2003 ',
 'Abc123456 ',
 'christina',
 'Owner123!@#',
 '012345',
 'qwert@123',
 'PaSswoRD',
 'PaSswoRD ',
 'P@ssw0rd$#@! ',
 'Qwerty@1234rfv',
 '1q@w3@e$r5',
 '123qweQWE.',
 '!@34qwera',
 'Pass_hash',
 'hh@kk123. ',
 'Passw0rd15 ',
 

In [38]:
failed_nonroot_countries = failed_nonroot.map(lambda line: get_country_name(line))
#failed_nonroot_countries.take(5)
failed_nonroot_countries.distinct().collect()

Traceback (most recent call last):
  File "/opt/spark-3.0.1-bin-hadoop3.2/python/pyspark/serializers.py", line 468, in dumps
    return cloudpickle.dumps(obj, pickle_protocol)
  File "/opt/spark-3.0.1-bin-hadoop3.2/python/pyspark/cloudpickle.py", line 1097, in dumps
    cp.dump(obj)
  File "/opt/spark-3.0.1-bin-hadoop3.2/python/pyspark/cloudpickle.py", line 357, in dump
    return Pickler.dump(self, obj)
  File "/usr/lib/python3.6/pickle.py", line 409, in dump
    self.save(obj)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 751, in save_tuple
    save(element)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/opt/spark-3.0.1-bin-hadoop3.2/python/pyspark/cloudpickle.py", line 501, in save_function
    self.save_function_tuple(obj)
  File "/opt/spark-3.0.1-bin-hadoop3.2/python/pyspark/cloudpickle.py", li

PicklingError: Could not serialize object: TypeError: cannot serialize '_io.BufferedReader' object

There are 76,541 failed attempts to access the server as non-root. A filter on the log files identifies all non-root entries, (explain sshd) with known usernames. All entries with a user name are filtered from failed_nonroot and assigned to failed_usernames. The values are then mapped according to the username in each entry using the get_username function. Each distinct value is outputted as list that identifies every username associated with failed login attempts.

## 3. Through out the durations of the log files, which date has the highest number of attack attempts? Anything interesting about that particular date?

In [ ]:
# Something to do with Chinese Labor Day (May 1)
dates = auth_log.map(lambda line: (re.split(pattern = r'\S{3}\W{1,2}\D{2}', string=line)[0], 1))
dates.take(5)
#countsByKey = dates.countByKey()
#countsByKey

The log files are mapped by the date which is formatted as one A-Z, two a-z, one or two whitespace characters, and two 0-9. The .countByKey() method outputs the number of attack attempts per date. The date with the most number of attacks is (date). This date is (explain).